drivers/sensor/microchip/mcp9843/mcp9843.c

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright (c) 2024 Vogl Electronic GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT microchip_mcp9843
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/kernel.h>
+#include <zephyr/logging/log.h>
+
+LOG_MODULE_REGISTER(MCP9843, CONFIG_SENSOR_LOG_LEVEL);
+
+#define MCP9843_REG_AMBIENT_TEMP 0x05
+#define MCP9843_REG_ID_REVISION  0x07
+#define MCP9843_REG_RESOLUTION   0x08
+
+struct mcp9843_data {
+	int32_t temp;
+};
+
+struct mcp9843_config {
+	const struct i2c_dt_spec bus;
+	uint8_t resolution;
+};
+
+static int mcp9843_reg_read(const struct device *dev, uint8_t start, uint8_t *buf, int size)
+{
+	const struct mcp9843_config *cfg = dev->config;
+
+	return i2c_burst_read_dt(&cfg->bus, start, buf, size);
+}
+
+static int mcp9843_set_temperature_resolution(const struct device *dev, uint8_t resolution)
+{
+	const struct mcp9843_config *cfg = dev->config;
+
+	return i2c_reg_write_byte_dt(&cfg->bus, MCP9843_REG_RESOLUTION, resolution);
+}
+
+static int mcp9843_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct mcp9843_data *data = dev->data;
+	uint8_t buf[2];
+	int ret;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_AMBIENT_TEMP) {
+		LOG_ERR("Unsupported sensor channel");
+		return -ENOTSUP;
+	}
+
+	ret = mcp9843_reg_read(dev, MCP9843_REG_AMBIENT_TEMP, buf, sizeof(buf));
+	if (ret < 0) {
+		data->temp = INT32_MIN;
+		return ret;
+	}
+
+	buf[0] &= 0x1F; /* mask out flags */
+
+	if ((buf[0] & 0x10) == 0x10) {
+		/* negative temperature */
+		buf[0] &= 0x0F;
+		data->temp = ((buf[0] * 16 * 16) + buf[1]) - (256 * 16);
+	} else {
+		/* positive temperature */
+		data->temp = (buf[0] * 16 * 16) + buf[1];
+	}
```

Comment:
Nits:

1. My brain kind of wants to see the conversion take place in a single place (preferrably during the `mcp9843_channel_get`) vs two partial conversions as it is now). This would make it easier to migrate into Sensor v2 model, where read and decode are clearly split.

2. Instead of handling byte-swapping with `((buf[0] * 16 * 16) + buf[1])`  I'd suggest: `sys_be16_to_cpu(masked_reg_value_u16)`

Reply:
You might want to do a proper sign extension before the `sys_be16_to_cpu` (i.e. `if(buf[0] & 0x10) {buf[0] |= 0xF0}`), since the number is already a signed int12 and only needs to be extended to 16 bits. That saves a few CPU cycles unless the compiler is smart enough to detect the intent of the original code.
N.b.: I am not aware of any sign extension function in Zephyr, but if there is any, its probably preferred to use that instead.